deribit-api/src/main/java/io/contek/invoker/deribit/api/websocket/WebSocketChannel.java

```diff
@@ -0,0 +1,90 @@
+package io.contek.invoker.deribit.api.websocket;
+
+import io.contek.invoker.commons.websocket.AnyWebSocketMessage;
+import io.contek.invoker.commons.websocket.BaseWebSocketChannel;
+import io.contek.invoker.commons.websocket.SubscriptionState;
+import io.contek.invoker.commons.websocket.WebSocketSession;
+import io.contek.invoker.deribit.api.websocket.common.WebSocketInboundMessage;
+import io.contek.invoker.deribit.api.websocket.common.WebSocketRequest;
+import io.contek.invoker.deribit.api.websocket.common.WebSocketResponse;
+import io.contek.invoker.deribit.api.websocket.common.constants.WebSocketOutboundKeys;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.Collections;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static io.contek.invoker.commons.websocket.SubscriptionState.*;
+
+@ThreadSafe
+public abstract class WebSocketChannel<Message extends WebSocketInboundMessage>
+  extends BaseWebSocketChannel<Message> {
+
+  private final AtomicReference<WebSocketRequest> pendingSubscriptionHolder = new AtomicReference<>();
+
+  protected abstract String getChannel();
+
+  @Override
+  protected final SubscriptionState subscribe(WebSocketSession session) {
+
+    synchronized (pendingSubscriptionHolder) {
+      if (pendingSubscriptionHolder.get() != null) {
+        throw new IllegalStateException();
+      }
+      WebSocketRequest request = new WebSocketRequest();
+      request.method = "public/" + WebSocketOutboundKeys._subscribe;
+      request.params.put("channels", Collections.singletonList(getChannel()));
```

Comment:
Always use Immutable(List/Set/Map) to be more consistent with the rest of the code base.